pkg/compact/compact.go

```diff
@@ -320,26 +360,143 @@ func (g *DefaultGrouper) Groups(blocks map[ulid.ULID]*metadata.Meta) (res []*Gro
+				g.enableVerticalCompaction,
+				g.enableConcurrentGroupCompaction,
+				g.compactions.WithLabelValues(groupKey),
+				g.compactionRunsStarted.WithLabelValues(groupKey),
+				g.compactionRunsCompleted.WithLabelValues(groupKey),
+				g.compactionFailures.WithLabelValues(groupKey),
+				g.verticalCompactions.WithLabelValues(groupKey),
+				g.garbageCollectedBlocks,
+				g.blocksMarkedForDeletion,
+				g.hashFunc,
+			)
+			if err != nil {
+				return nil, errors.Wrap(err, "create compaction group")
+			}
+			groups[groupKey] = group
+			res = append(res, group)
+		}
+		if err := group.AppendMeta(m); err != nil {
+			return nil, errors.Wrap(err, "add compaction group")
+		}
+	}
+
+	newGroups := []*Group{}
+	// Split each group by available plans.
+	for _, group := range res {
+		newGroup, err := g.splitGroupByPlans(group)
+		if err != nil {
+			return nil, err
+		}
+		newGroups = append(newGroups, newGroup...)
+	}
+	res = newGroups
+
+	sort.Slice(res, func(i, j int) bool {
+		return res[i].Key() < res[j].Key()
+	})
+	return res, nil
+}
+
+// Recursively plan a groups metas until there are no available plans and return a new group for each plan.
+// This allows a single group key to utilize group concurrency when there are multiple plans available to compact.
+func (g *DefaultGrouper) splitGroupByPlans(group *Group) ([]*Group, error) {
+	groups := []*Group{}
+	for {
+		var toCompact []*metadata.Meta
+		var err error
+		if len(group.metasByMinTime) != 0 {
+			toCompact, err = g.planner.Plan(context.Background(), group.metasByMinTime)
+		}
+		if err != nil {
+			return nil, errors.Wrap(err, "split group")
+		}
+		if len(groups) > 0 && len(toCompact) == 0 {
+			break
+		}
+		lbls := group.Labels()
+		groupKey := group.Key()
+		newGroup, err := NewGroup(
+			log.With(g.logger, "split group", fmt.Sprintf("%d@%v", group.Resolution(), lbls.String()), "groupKey", groupKey),
+			g.bkt,
+			groupKey,
+			lbls,
+			group.Resolution(),
+			group.acceptMalformedIndex,
+			g.enableVerticalCompaction,
+			g.enableConcurrentGroupCompaction,
+			g.compactions.WithLabelValues(groupKey),
+			g.compactionRunsStarted.WithLabelValues(groupKey),
+			g.compactionRunsCompleted.WithLabelValues(groupKey),
+			g.compactionFailures.WithLabelValues(groupKey),
+			g.verticalCompactions.WithLabelValues(groupKey),
+			g.garbageCollectedBlocks,
+			g.blocksMarkedForDeletion,
+			g.hashFunc,
+		)
+		if err != nil {
+			return nil, errors.Wrap(err, "create split compaction group")
+		}
+		groups = append(groups, newGroup)
+		newMetasByMinTime := []*metadata.Meta{}
+		ids := map[ulid.ULID]bool{}
+		for _, m := range toCompact {
+			if err := newGroup.AppendMeta(m); err != nil {
+				return nil, errors.Wrap(err, "add to split compaction group")
+			}
+			ids[m.ULID] = true
+		}
+		for _, m := range group.metasByMinTime {
+			if _, ok := ids[m.ULID]; !ok {
+				newMetasByMinTime = append(newMetasByMinTime, m)
+			}
+		}
+		group.metasByMinTime = newMetasByMinTime
```

Comment:
Ok, this is something I am the most worried about.

I think this algorithm might be quite ineffective in practice. The reason is that compaction usually depends on each other. I would need to double check but in my experience default compaction planner is pretty naive.

For example, if it sees let's say

```
[2d][8h][8h][2h][2h][2h][2h][2h]
```

First plan would say `[2d][8h][8h]  (let me compact those [2h][2h][2h][2h])   [2h]`

If you take them out, compact will assume gap and will compact concurrently  `[2d][8h][8h] <2d gap>  [2h]` into 2d block:  `[2d] ( let me compact those [8h][8h] <2d gap> ) [2h]`. This is because the compaction algorithm was designed for single group Prometheus one by one blocks.

Consequences for Cortex are lower: You just recompact multiple times unnecessarily. For Thanos for deployments that does not have vertical compaction enabled it will just fail compactor (overlapping blocks).

That's why I believe we have to look at a smarter planner that will give us the planned compactions that we can concurrently attack and will be able to give us feedback anytime (even if other compactions are running) on what to run.

Feel free to change the majority of the interface to accomplish this well with Cortex. We are happy in Thanos to default to the concurrent solution (with simply worker numbers = 1). So I wonder if the current grouper/planner/compaction fits here. Free to rearrange those 3 APIs from ground zero. It's better than just fiddling with something not fitting on existing not well prepared API for concurrency on this level.